lib/galaxy/datatypes/data.py

```diff
@@ -690,8 +690,21 @@ def matches_any(self, target_datatypes):
         Check if this datatype is of any of the target_datatypes or is
         a subtype thereof.
         """
+        self_datatype = self
+
+        if hasattr(self, "uncompressed_datatype_class"):
+            # Compare using uncompressed datatype instead of self and filter comparisons
+            # to only uncompressed datatypes corresponding to supplied compressed datatypes.
+            potential_uncompressed_types = []
+            for target_datatype in target_datatypes:
+                if hasattr(target_datatype, "uncompressed_datatype_class") and target_datatype.compressed_format == self.compressed_format:
+                    potential_uncompressed_types.append(target_datatype.uncompressed_datatype_class)
+
+            self_datatype = self.uncompressed_datatype_class
+            target_datatypes = potential_uncompressed_types
+
         datatype_classes = tuple(datatype if isclass(datatype) else datatype.__class__ for datatype in target_datatypes)
```

Comment:
This?

```diff
diff --git a/lib/galaxy/datatypes/data.py b/lib/galaxy/datatypes/data.py
index 1734312c79..40e5aeabdd 100644
--- a/lib/galaxy/datatypes/data.py
+++ b/lib/galaxy/datatypes/data.py
@@ -692,6 +692,7 @@ class Data(object):
         """
         self_datatype = self

+        target_datatypes = tuple(datatype if isclass(datatype) else datatype.__class__ for datatype in target_datatype
         if hasattr(self, "uncompressed_datatype_class"):
             # Compare using uncompressed datatype instead of self and filter comparisons
             # to only uncompressed datatypes corresponding to supplied compressed datatypes.
@@ -701,10 +702,9 @@ class Data(object):
                     potential_uncompressed_types.append(target_datatype.uncompressed_datatype_class)

             self_datatype = self.uncompressed_datatype_class
-            target_datatypes = potential_uncompressed_types
+            target_datatypes = tuple(potential_uncompressed_types)

-        datatype_classes = tuple(datatype if isclass(datatype) else datatype.__class__ for datatype in target_datatype
-        return isinstance(self_datatype, datatype_classes)
+        return isinstance(self_datatype, target_datatypes)

     def merge(split_files, output_file):
         """
```

I can commit that.